### PR TITLE
tpm2_print: A tool to print TPM data structures

### PR DESCRIPTION
**Pull request by harisokanovic**
Takes an object type + file path as inputs and prints enclosed elements
to stdout in YAML.

Presently only supports TPMS_ATTEST objects for the purpose of printing
TPM generated quotes. Does NOT authenticate input file, which can be
done using other tools -- E.g. see edc3e022 for openssl example.

### CODE REVIEW

**Comment by martinezjavier** [comment]
@harisokanovic can you please fix the compile error and resend?

```
 CC       tools/aux/tpm2_print.o
../tools/aux/tpm2_print.c: In function ‘print_TPMS_QUOTE_INFO’:
../tools/aux/tpm2_print.c:119:5: error: ‘for’ loop initial declarations are only allowed in C99 mode
     for(long unsigned int i = 0; i < pcr_selection_count; ++i) {
     ^
../tools/aux/tpm2_print.c:119:5: note: use option -std=c99 or -std=gnu99 to compile your code
make[1]: *** [tools/aux/tpm2_print.o] Error 1
make[1]: Leaving directory `/home/travis/build/intel/tpm2-tools/tpm2-tools-3.0.1-54-gb286bf3/_build'
make: *** [distcheck] Error 1
```

File: lib/tpm2_util.c [hunk 1]
```diff
@@ -132,42 +132,71 @@ int tpm2_util_hex_to_byte_structure(const char *inStr, UINT16 *byteLength,
     return 0;
 }
 
-void tpm2_util_hexdump(BYTE *data, size_t len, bool plain) {
+void tpm2_util_hexdump(const BYTE *data, size_t len, bool plain) {
+    FILE *fd = fmemopen((void*)data, len, "r");
+    if (!fd) {
+        /* TODO: should we change the API to allow error? */
```
Comment (martinezjavier) [inline lib/tpm2_util.c]: AFAICT, the only caller of this function currently is tpm2_util_public_to_yaml(), so it should be quite easy to change the function signature to propagate the error.

That said, I wonder if instead of using fmemopen() and calling tpm2_util_hexdump_file(), can we do the opposite and make tpm2_util_hexdump_file() read the file contents and call tpm2_util_hexdump()?

Since once the data is in memory, tpm2_util_hexdump() shouldn't really have an error path.

File: lib/tpm2_util.h [hunk 2]
```diff
@@ -155,7 +156,24 @@ bool tpm2_util_string_to_uint16(const char *str, uint16_t *value);
...
+/**
+ * Prints an xxd compatible hexdump to stdout if output is enabled,
+ * ie no -Q option.
+ *
+ * @param fd
+ *  A readable open file.
+ * @param len
+ *  The length of the data to read and print.
+ * @param plain
+ *  true for a plain hex string false for an xxd compatable
```
Comment (martinezjavier) [inline lib/tpm2_util.h]: s/compatable/compatible

File: Makefile.am [hunk 3]
```diff
@@ -48,6 +48,7 @@ LDADD = \
 
 # keep me sorted
 bin_PROGRAMS = \
+    tools/aux/tpm2_print \
```
Comment (martinezjavier) [inline Makefile.am]: Did we decide that there would be an aux sub-directory? I remember the discussion but I had that at the end we decided to just have all the tools under tools.

File: man/tpm2_print.1.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+% tpm2_print(1) tpm2-tools | General Commands Manual
+%
+% DECEMBER 2018
```
Comment (martinezjavier) [inline man/tpm2_print.1.md]: I don't think this patch will take a year to be merged :smiley:

File: tools/aux/tpm2_print.c [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdio.h>
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: Please insert a blank line here to separate system headers from project headers.

File: tools/aux/tpm2_print.c [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+#include "log.h"
+#include "files.h"
+#include "tpm2_util.h"
+#include "tpm2_alg_util.h"
+#include "tpm2_tool.h"
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: Please sort these alphabetically.

File: tools/aux/tpm2_print.c [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+const char* file_path = "";
+
+#define FILE_2_SHORT_ERROR() LOG_ERR("%s: File too short", (file_path))
+
+static inline void print_yaml_indent(size_t indent_count)
+{
+    while (indent_count-- > 0) {
+        tpm2_tool_output("  ");
+    }
+}
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: This probably should be in lib/tpm2_util.{c,h}

File: tools/aux/tpm2_print.c [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+
+typedef enum {
+    file_type_unknown = 0,
+    file_type_TPMS_ATTEST
+} file_type_id;
+
+file_type_id file_type = file_type_unknown;
+const char* file_path = "";
+
+#define FILE_2_SHORT_ERROR() LOG_ERR("%s: File too short", (file_path))
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: This is only used in two places, so I wonder the value of the macro. But it's probably just my personal taste, since I think that is an unnecessary level of indirection and makes the code harder to read.

File: tools/aux/tpm2_print.c [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+    tpm2_tool_output("restartCount: %lu\n", (long unsigned int)numb.u32);
+
+    res = files_read_bytes(fd, &numb.u8, 1);
+    if (!res) {
+        goto read_error;
+    }
+    print_yaml_indent(indent_count);
+    tpm2_tool_output("safe: %u\n", (unsigned int)numb.u8);
+
+    // success
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: I think this comment is superfluous.

File: tools/aux/tpm2_print.c [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    tpm2_tool_output("pcrSelect:\n");
+
+    print_yaml_indent(indent_count + 1);
+    tpm2_tool_output("count: %lu\n", (long unsigned int)pcr_selection_count);
+
+    print_yaml_indent(indent_count + 1);
+    tpm2_tool_output("pcrSelections:\n");
+
+    // read TPML_PCR_SELECTION array (of size count)
+    for(long unsigned int i = 0; i < pcr_selection_count; ++i) {
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: This `i` declaration is what should be fixed.

File: tools/aux/tpm2_print.c [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+    };
+
+    *opts = tpm2_options_new("t:f:", ARRAY_LEN(topts), topts,
+        on_option, NULL);
+
+    return *opts != NULL;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+    UNUSED(sapi_context);
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: Not related with your patch, but this tool could be used even if there's no TPM available. But currently it will fail even when the SAPI context is not used. I've filled issue #737 for that.

File: tools/aux/tpm2_print.c [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+    UNUSED(flags);
+
+    bool res;
+    FILE* fd = stdin;
+
+    if (file_path && *file_path != 0) {
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: Maybe change to `if (file_path && *file_path)` for consistency?

File: tools/aux/tpm2_print.c [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+            goto error;
+        }
+        break;
+    default:
+        LOG_ERR("Invalid file_type");
+        goto error;
+        break;
+    }
+
+    // log number of bytes read
```
Comment (martinezjavier) [inline tools/aux/tpm2_print.c]: Again, this and the following `// success` comments could be removed IMHO.

File: Makefile.am (hunk 3 above)
Comment (williamcroberts) [reply Makefile.am]: I don't think we ever concluded, I'm OK with an aux tool directory, what about you @martinezjavier ?

File: lib/tpm2_util.c (hunk 1 above)
Comment (williamcroberts) [reply lib/tpm2_util.c]: I think I prefer it this way, as potentially big buffers won't need to be fully read into memory. Granted, most tpm data structures are quite small. Either way is fine by me in the end.

File: tools/aux/tpm2_print.c (hunk 11 above)
Comment (williamcroberts) [reply tools/aux/tpm2_print.c]: Yeah perhaps we need a way in onstart to indicate that they don't need a SAPI. tpm2_rc_decode has the same issue, and could go to aux as well at some point.

File: tools/aux/tpm2_print.c (hunk 12 above)
Comment (williamcroberts) [reply tools/aux/tpm2_print.c]: fopen("", "rb) will fail, don't even need that check.

File: tools/aux/tpm2_print.c [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+read_error:
+    FILE_2_SHORT_ERROR();
+error:
+    return false;
+}
+
+static bool on_option(char key, char *value) {
+    switch (key) {
+    case 't':
+        if (strcmp(value, "TPMS_ATTEST") == 0) {
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: Flipping the logic makes this simpler. If not TPMS_ATTEST, log an error and return, no need for an else clause, just assign to type.
```
case 't':
    if (strcmp(value, "TPMS_ATTEST"))  {
        LOG_ERR("msg"):
        return 1;
    }
    file_type = file_type_TPMS_ATTEST;
    break;
```

File: tools/aux/tpm2_print.c [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+    // success
+    return true;
+
+read_error:
+    FILE_2_SHORT_ERROR();
+error:
+    return false;
+}
+
+static bool print_TPMS_ATTEST_yaml(FILE* fd, size_t indent_count) {
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: Whats the point if indent_count, all I see is it being set to 0, subordinate calls are settiing it to other values, so I see no point in setting it here.

File: tools/aux/tpm2_print.c [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+    bool res;
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: declare and assign at first use on 196 per the coding standard.

File: tools/aux/tpm2_print.c [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+} file_type_id;
+
+file_type_id file_type = file_type_unknown;
+const char* file_path = "";
+
+#define FILE_2_SHORT_ERROR() LOG_ERR("%s: File too short", (file_path))
+
+static inline void print_yaml_indent(size_t indent_count)
+{
+    while (indent_count-- > 0) {
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: no need for ```> 0``` when indent_count is 0, it's false and will stop.

File: tools/aux/tpm2_print.c (hunk 10 above)
Comment (williamcroberts) [reply tools/aux/tpm2_print.c]: Yes keep types consistent, I know their are some sketchy spots in this codebase, but Id like to not add more.

File: man/tpm2_print.1.md [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+**tpm2_print** [*OPTIONS*]
+
+# DESCRIPTION
+
+**tpm2_print**(1) decodes a TPM data structure and prints enclosed
+elements to stdout as YAML.
+
+# OPTIONS
+
+  * **-t**, **--type**:
```
Comment (williamcroberts) [inline man/tpm2_print.1.md]: Is there anyway we could figure this out for the user? I didn't see anything as a dead giveaway...

File: man/tpm2_print.1.md (hunk 4 above)
Comment (harisokanovic) [reply man/tpm2_print.1.md]: :) fixed

File: tools/aux/tpm2_print.c (hunk 5 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Added

File: tools/aux/tpm2_print.c (hunk 6 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: done

File: tools/aux/tpm2_print.c (hunk 7 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Moved

File: tools/aux/tpm2_print.c (hunk 8 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Alright, I removed the macro since both you and William don't like it :)

File: tools/aux/tpm2_print.c (hunk 9 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Removed

File: tools/aux/tpm2_print.c (hunk 10 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Fixed. I guess my compiler defaults to C99 and missed this.

Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: I used `long unsigned int` to avoid extra casting on prints below. Longs are guaranteed to be at least 32 bits wide, but I can change it if you prefer.

File: tools/aux/tpm2_print.c (hunk 11 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: I noticed that as well, but it hasn't been an issue for us yet. I'd prefer to compile out the SAPI initialization instead of a runtime flag, so that we can remove the library dependency as well.

File: tools/aux/tpm2_print.c (hunk 12 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: `file_path` defaults to empty string, so we need `*file_path`.

File: tools/aux/tpm2_print.c (hunk 13 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Removed

File: tools/aux/tpm2_print.c (hunk 17 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Removed

File: tools/aux/tpm2_print.c (hunk 15 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: I was just being consistent with the other print_*() functions. I can remove it.

File: tools/aux/tpm2_print.c (hunk 16 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Fixed

File: tools/aux/tpm2_print.c (hunk 14 above)
Comment (harisokanovic) [reply tools/aux/tpm2_print.c]: Changed

File: lib/tpm2_util.h (hunk 2 above)
Comment (harisokanovic) [reply lib/tpm2_util.h]: Fixed

File: Makefile.am (hunk 3 above)
Comment (martinezjavier) [reply Makefile.am]: @williamcroberts I'm also OK with it. I just asked since it was not clear to me that there was an agreement on that.

File: tools/aux/tpm2_print.c (hunk 11 above)
Comment (williamcroberts) [reply tools/aux/tpm2_print.c]: @harisokanovic you'd still need the headers for data types, so it would only remove the link dependency.

File: tools/aux/tpm2_print.c [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+    UNUSED(sapi_context);
+    UNUSED(flags);
+
+    bool res;
+    FILE* fd = stdin;
+
+    if (file_path && *file_path) {
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: NULL file_path, no need for it to be "".
If NULL set file_path to something like "<stdin>" in here.

File: tools/aux/tpm2_print.c [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+    default:
+        LOG_ERR("Invalid file_type");
+        goto error;
+        break;
+    }
+
+    // log number of bytes read
+    LOG_INFO("Read %ld bytes from file_path=%s", ftell(fd), file_path);
+
+    fclose(fd);
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: looks like your closing stdin, you need to check if (fd == stdin) don't fclose it.

File: tools/aux/tpm2_print.c [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+    case file_type_TPMS_ATTEST:
+        res = print_TPMS_ATTEST_yaml(fd);
+        if(!res) {
+            goto error;
+        }
+        break;
+    default:
+        LOG_ERR("Invalid file_type");
+        goto error;
+        break;
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: no need for break after goto, unless it silences a silly compiler warning.

File: tools/aux/tpm2_print.c [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+    }
+
+    // log number of bytes read
+    LOG_INFO("Read %ld bytes from file_path=%s", ftell(fd), file_path);
+
+    fclose(fd);
+    return 0;
+
+error:
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: you can just put an out label and return an rc variable. Less duplication of cleanup code.
```
function() {
    int rc = 1;
    do_stuff()
     if (fail) {
          goto out;
     }
    rc = 0;

out:
     cleanup();
     return rc;
```

File: tools/aux/tpm2_print.c [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+            LOG_ERR("%s: Could not open file", file_path);
+            return 1;
+        }
+    }
+    else {
+        LOG_INFO("Reading from stdin");
+    }
+
+    switch (file_type) {
+    case file_type_TPMS_ATTEST:
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: You could check this before even attempting to read the file.... this way you can fail early before you even have to worry about closing the file.

File: tools/aux/tpm2_print.c [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef enum {
+    file_type_unknown = 0,
+    file_type_TPMS_ATTEST
+} file_type_id;
+
+file_type_id file_type = file_type_unknown;
+const char* file_path = "";
```
Comment (williamcroberts) [inline tools/aux/tpm2_print.c]: NULL makes more sense here, also static for both vars.

Also, follow convention and create a context structure for these...
```
typedef struct tpm2_print_ctx tpm2_print_ctx;
struct tpm2_print_ctx {
  struct {
    const char *path;
    file_type_id type;
  } file;
};

static tpm2_print_ctx ctx;

foo() {
  ctx.file.path = "foo";
  ctx.file.type = 43;
}
```

File: tools/aux/tpm2_print.c (hunk 19 above)
Comment (williamcroberts) [reply tools/aux/tpm2_print.c]: If you do it that way, you can also use https://github.com/intel/tpm2-tools/blob/master/lib/files.h#L67:
```
/**
 * Loads data from a file path or stdin enforcing an upper bound on size.
 * @param path
 *  The path to load data from, NULL means stdin.
 * @param size
 *  The maximum size.
 * @param buf
 *  The buffer to write the data into.
 * @return
 *  True on success or false otherwise.
 */
bool files_load_bytes_from_file_or_stdin(const char *path, UINT16 *size, BYTE *buf);
```

File: man/tpm2_print.1.md (hunk 18 above)
Comment (harisokanovic) [reply man/tpm2_print.1.md]: We could do something like assume type=TPMS_ATTEST when the first 4 bytes (magic) are ff544347, but that's a bit flaky. I'm also not sure that every TPM generated object uses the same magic. We can always add this later without breaking the API.

**Comment by harisokanovic** [comment]
@williamcroberts , @martinezjavier , I posted an update with changes you requested + a first attempt at splitting TCTI and SAPI initialization from the arg parser, so that we can avoid it in aux tools. Looks like travis-ci is hitting a build error on an unrelated file. Any ideas?

**Comment by williamcroberts** [comment]
@harisokanovic Thank you for your patience and update to the PR. I will look to make sure that master is still building, offhand, looks like somethings changed in the TSS so master may be broken on travis. In the meantime, if you can rebase your change on the latest and fix the conflicts, that would be great.

**Comment by harisokanovic** [comment]
@williamcroberts, looks like the travis-ci build succeeded

File: lib/tpm2_options.h [hunk 25]
```diff
@@ -45,8 +45,12 @@ union tpm2_option_flags {
         UINT8 verbose : 1;
         UINT8 quiet   : 1;
         UINT8 enable_errata  : 1;
+        struct {
+            char *name;
+            char *opts;
+        } tcti;
     };
-    UINT8 all;
+    UINT8 all : 3;
```
Comment (williamcroberts) [inline lib/tpm2_options.h]: Why did you change this?

Comment (harisokanovic) [reply lib/tpm2_options.h]: I moved sapi initialization to avoid a run-time dependency on libsapi. However, it doesn't matter in my use case (yet anyway), so I could go either way on this point.

I added an explicit 3-bit deceleration for `all`, since the rest of the bits in this data structure are char*'s. I agree it's a bit weird to add strings to the flags struct. I can move it if there's a better/preferred place.

Comment (williamcroberts) [reply lib/tpm2_options.h]: The three bit bitfield on all, kill it off. It will likely generate more code as the compiler cant just write 0 to a byte address, it would have to mask off and clear only those bits... ie read a byte, and them with 0x7 and write it back.

For the sapi split, I am not worried about the link dependency, so you'll have to change that up to the tpm2_options_new flag approach.

Otherwise, once those are done, ill be happy to pull this in.

Comment (williamcroberts) [reply lib/tpm2_options.h]: FYI, you can use: https://github.com/intel/tpm2-tools/pull/767 as a sample to port to master. This may or may not be done by the time you get back to this.

**Comment by harisokanovic** [comment]
@williamcroberts , I replaced the SAPI split with a "flags" bitmask in tpm2_options_new() as you suggested.

**Comment by harisokanovic** [comment]
Great, thanks. I forgot about the man pages. We should also move `tpm2_rc_decode` into the aux/ directory and apply the NO_SAPI flag.

**Comment by williamcroberts** [comment]
See #769 for existing work items
